### DISCUSSION TITLE
v7.1.0-30 bug in FX expression

### DISCUSSION THREAD

**Discussion by GeeMack** [ImageMagick/ImageMagick#5066] [Development] [answered]
Using ImageMagick 7.1.0-30 Q16-HDRI on Windows 10, I've encountered an error new since the previous version.

This command should print the result of the FX expression as "70x46", but prints "0x0" instead...

    magick rose: -print "%[fx:u.w]x%[fx:u.h]\n" null:

This command should output the "rose;" and the "logo:" image scaled to the same size, but instead it prints an error and ends the command...

    magick rose: logo: -scale %[fx:u.w]x%[fx:u.h]! test.png
    magick: negative or zero image size `ROSE' @ error/resize.c/ScaleImage/4116.

Where the expression `%[fx:w]` will return the width, this `%[fx:u.w]` returns zero, which is pretty obviously not the desired behavior. Specifying an image from the list using any construct like `u[2]` or `v` or `u[-2]` seems to fail.

**Comment by snibgo** [comment]
Eek! Thanks for the problem report.

Please try the following:
```
magick rose: -define fx:debug=true -format %[fx:w] info:
70

magick rose: -define fx:debug=true -format %[fx:u.w] info:
70
```
... and paste the text output here.

With my development version of IM, v7.1.0-20, I can't reproduce the problem:
```
f:\web\im>%FXNEW%magick rose: -print "%[fx:u.w]x%[fx:u.h]\n" null:
70x46

f:\web\im>%FXNEW%magick rose: logo: -scale %[fx:u.w]x%[fx:u.h]! test.png

f:\web\im>%FXNEW%magick test.png info:
test.png PNG 93x58 93x58+0+0 8-bit Gray 1366B 0.859u 0:00.902
```

I'll need to download v7.1.0-30, both pre-built and source, and see what is going on. And I'm tied up for a few days, so expect this will take a week or so.

EDIT to add: I haven't changed fx.c since about v7.1.0-21.

**Reply by GeeMack** [reply]
The results of the first command...

    magick rose: -define fx:debug=true -format %[fx:w] info:
    Operators:
      += -= *= /= ++ -- + - * / % + - << >> == != <= >= < > && || ! & | ~ ^ ? : ( ) [ ] { } =
    Functions:
       abs acos airy alt asin atan2 atan ceil channel clamp cosh cos debug drc erf exp floor gauss gcd hypot int isnan j0 j1 jinc ln logtwo log max min mod not pow rand round sign sinc sinh sin sqrt squish tanh tan trunc do for if while u u0 up s v p sp vp
    Image attributes:
       depth extent kurtosis maxima mean median minima page page.x page.y page.width page.height printsize printsize.x printsize.y quality resolution resolution.x resolution.y skewness standard_deviation h n t w z
    Symbols:
       hue intensity lightness luma luminance saturation a b c g i j k m o r y
    Controls:
       goto ifzerogoto ifnotzerogoto copyfrom copyto zerstk 
    UserSymbols (0)
    DumpRPN:  numElements=100  usedElements=1  maxUsedOprStack=0  ImgListLen=1  NeedStats=no  GotStats=no  NeedHsl=no
    CornerOnly
      0: ImgAttr val=0 'w' nArgs=0 ndx=0  push
    70

The results of the second  command...

    magick rose: -define fx:debug=true -format %[fx:u.w] info:    Operators:
      += -= *= /= ++ -- + - * / % + - << >> == != <= >= < > && || ! & | ~ ^ ? : ( ) [ ] { } =
    Functions:
       abs acos airy alt asin atan2 atan ceil channel clamp cosh cos debug drc erf exp floor gauss gcd hypot int isnan j0 j1 jinc ln logtwo log max min mod not pow rand round sign sinc sinh sin sqrt squish tanh tan trunc do for if while u u0 up s v p sp vp
    Image attributes:
       depth extent kurtosis maxima mean median minima page page.x page.y page.width page.height printsize printsize.x printsize.y quality resolution resolution.x resolution.y skewness standard_deviation h n t w z
    Symbols:
       hue intensity lightness luma luminance saturation a b c g i j k m o r y
    Controls:
       goto ifzerogoto ifnotzerogoto copyfrom copyto zerstk 
    UserSymbols (0)
    DumpRPN:  numElements=100  usedElements=2  maxUsedOprStack=1  ImgListLen=1  NeedStats=no  GotStats=no  NeedHsl=no
    CornerOnly
      0: Constant val=0 'onull' nArgs=0 ndx=0  push
      1: Function val=0 'u' nArgs=1 ndx=0  push ia=w
    0

**Comment by fmw42** [comment]
Gives the wrong answer on my Mac OSX Sierra with

```
magick -version
Version: ImageMagick 7.1.0-30 Q16-HDRI x86_64 19924 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules 
Delegates (built-in): bzlib cairo djvu fftw fontconfig freetype gslib gvc heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr png ps raw rsvg tiff webp x xml zlib
Compiler: gcc (4.2)

```

```
magick rose: -define fx:debug=true -format %[fx:w] info:
Operators:
  += -= *= /= ++ -- + - * / % + - << >> == != <= >= < > && || ! & | ~ ^ ? : ( ) [ ] { } =
Functions:
   abs acosh acos airy alt asinh asin atanh atan2 atan ceil channel clamp cosh cos debug drc erf exp floor gauss gcd hypot int isnan j0 j1 jinc ln logtwo log max min mod not pow rand round sign sinc sinh sin sqrt squish tanh tan trunc do for if while u u0 up s v p sp vp
Image attributes:
   depth extent kurtosis maxima mean median minima page page.x page.y page.width page.height printsize printsize.x printsize.y quality resolution resolution.x resolution.y skewness standard_deviation h n t w z
Symbols:
   hue intensity lightness luma luminance saturation a b c g i j k m o r y
Controls:
   goto ifzerogoto ifnotzerogoto copyfrom copyto zerstk 
UserSymbols (0)
DumpRPN:  numElements=100  usedElements=1  maxUsedOprStack=0  ImgListLen=1  NeedStats=no  GotStats=no  NeedHsl=no
CornerOnly
  0: ImgAttr val=0 'w' nArgs=0 ndx=0  push
70
```

and

```
magick rose: -define fx:debug=true -format %[fx:u.w] info:
Operators:
  += -= *= /= ++ -- + - * / % + - << >> == != <= >= < > && || ! & | ~ ^ ? : ( ) [ ] { } =
Functions:
   abs acosh acos airy alt asinh asin atanh atan2 atan ceil channel clamp cosh cos debug drc erf exp floor gauss gcd hypot int isnan j0 j1 jinc ln logtwo log max min mod not pow rand round sign sinc sinh sin sqrt squish tanh tan trunc do for if while u u0 up s v p sp vp
Image attributes:
   depth extent kurtosis maxima mean median minima page page.x page.y page.width page.height printsize printsize.x printsize.y quality resolution resolution.x resolution.y skewness standard_deviation h n t w z
Symbols:
   hue intensity lightness luma luminance saturation a b c g i j k m o r y
Controls:
   goto ifzerogoto ifnotzerogoto copyfrom copyto zerstk 
UserSymbols (0)
DumpRPN:  numElements=100  usedElements=2  maxUsedOprStack=1  ImgListLen=1  NeedStats=no  GotStats=no  NeedHsl=no
CornerOnly
  0: Constant val=0 'onull' nArgs=0 ndx=0  push
  1: Function val=0 'u' nArgs=1 ndx=0  push ia=w
0
```

**Comment by urban-warrior** [comment, marked as answer]
Thanks for the problem report.  We can reproduce it and will have a patch to fix it in the GIT main branch @ https://github.com/ImageMagick/ImageMagick later today.  The patch will be available in the beta releases of ImageMagick @ https://imagemagick.org/download/beta/ by sometime tomorrow.

**Comment by snibgo** [comment]
Thanks, both. The debug shows that the translation phase of fx is working fine.

A Wizard seems to have already fixed the problem. Unless told otherwise, I'll assume this is fixed.

**Comment by GeeMack** [comment]
This issue appears to be fixed in release v7.1.0-31.
